File: src/main/java/com/jcloisterzone/figure/Mayor.java
package com.jcloisterzone.figure;

import com.jcloisterzone.Immutable;
import com.jcloisterzone.Player;
import com.jcloisterzone.board.Location;
import com.jcloisterzone.board.pointer.FeaturePointer;
import com.jcloisterzone.feature.*;
import com.jcloisterzone.game.state.GameState;

@Immutable
public class Mayor extends Follower {

    private static final long serialVersionUID = 1L;

    public Mayor(String id, Player player) {
        super(id, player);
    }

    @Override
    public int getPower(GameState state, Scoreable feature) {
        if (feature instanceof City) {
            if (((City) feature).hasModifier(state, City.ELIMINATED_PENNANTS)) {
            	return 0; // Pennants are eliminated in the City
            }
            return ((City) feature).getModifier(state, City.PENNANTS, 0);
        } else {
            // If not City, then it must be Castle. Mayor has no power on Castles
            return 0;
        }
    }

    @Override
    public DeploymentCheckResult isDeploymentAllowed(GameState state, FeaturePointer fp, Structure feature) {
        if (!(fp.getLocation() == Location.QUARTER_CASTLE ||
              feature instanceof City || feature instanceof FlyingMachine|| feature instanceof TrapFeature)) {
            return new DeploymentCheckResult("Mayor must be placed in city only.");
        }
        return super.isDeploymentAllowed(state, fp, feature);
    }

}

File: src/main/ts/com/jcloisterzone/figure/Mayor.ts
import type { Player } from "../Player.js";
import { Location } from "../board/Location.js";
import type { FeaturePointer } from "../board/pointer/FeaturePointer.js";
import { City } from "../feature/City.js";
import { FlyingMachine } from "../feature/FlyingMachine.js";
import type { Scoreable } from "../feature/Scoreable.js";
import type { Structure } from "../feature/Structure.js";
import { isInstanceOfTrapFeature } from "../feature/TrapFeature.js";
import type { GameState } from "../game/state/GameState.js";
import { DeploymentCheckResult } from "./DeploymentCheckResult.js";
import { Follower } from "./Follower.js";

/** The Mayor (Count, King &amp; Robber) — power = city pennants. */
export class Mayor extends Follower {
  static readonly simpleName = "Mayor";

  constructor(id: string, player: Player) {
    super(id, player);
  }

  override getPower(state: GameState, feature: Scoreable): number {
    if (feature instanceof City) {
      if (feature.hasModifier(state, City.ELIMINATED_PENNANTS)) {
        return 0;
      }
      return feature.getModifier(state, City.PENNANTS, 0);
    }
    return 0; // not a City -> Castle, Mayor has no power
  }

  override isDeploymentAllowed(state: GameState, fp: FeaturePointer, feature: Structure): DeploymentCheckResult {
    if (
      !(
        fp.getLocation() === Location.QUARTER_CASTLE ||
        feature instanceof City ||
        feature instanceof FlyingMachine ||
        isInstanceOfTrapFeature(feature)
      )
    ) {
      return new DeploymentCheckResult("Mayor must be placed in city only.");
    }
    return super.isDeploymentAllowed(state, fp, feature);
  }
}
